In [18]:
from pathlib import Path
import pandas as pd
import numpy as np
import json
from functools import reduce
from sklearn.preprocessing import MinMaxScaler

In [19]:
pd.set_option('display.max_columns', 1000)

In [20]:
# Defining directories
dir_path = Path.cwd().parent
data_folder = dir_path.joinpath("data/processed")

#training frames
s1_path = data_folder.joinpath("sent1_train_season_medians_widened.parquet")
s2_path = data_folder.joinpath("sent2_train_season_medians_widened.parquet")
cov_path = data_folder.joinpath("cov_final.parquet")

model_df_path = data_folder.joinpath("model_df.parquet")
# species_model_df_path = data_folder.joinpath('species_train_df.csv')
species_dict_path = dir_path.joinpath("data/species_dict.json")

index_cols = ["long", "lat", 'year',"species_names"]


# previous model frame path 
actual_model_path = dir_path.joinpath("data/processed_532_features/species_train_df.csv")

In [21]:
s1 = pd.read_parquet(s1_path)
s1

long                 lat  year          species_names  \
0       75.66669580000007   32.25107380000003  2018      Bambusoideae spp.   
1       75.66669580000007   32.25107380000003  2019      Bambusoideae spp.   
2       75.66669580000007   32.25107380000003  2020      Bambusoideae spp.   
3       75.66669580000007   32.25107380000003  2021      Bambusoideae spp.   
4       75.66669580000007   32.25107380000003  2022      Bambusoideae spp.   
...                   ...                 ...   ...                    ...   
893437  78.12624330000006  31.258713300000068  2019  Rhododendron arboreum   
893438  78.12624330000006  31.258713300000068  2020  Rhododendron arboreum   
893439  78.12624330000006  31.258713300000068  2021  Rhododendron arboreum   
893440  78.12624330000006  31.258713300000068  2022  Rhododendron arboreum   
893441  78.12624330000006  31.258713300000068  2023  Rhododendron arboreum   

        vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                 22.552725            23.639657            19.700091   
1                 23.856063            23.135292            19.956352   
2                 23.268281            23.263359            21.227070   
3                 24.927604            24.248816            19.777236   
4                 24.149733            22.787536            20.453230   
...                     ...                  ...                  ...   
893437            24.899585            21.450173            22.922561   
893438            21.172713            21.895621            18.450980   
893439            20.863598            23.242825            19.138139   
893440            20.043214            20.606978            17.781513   
893441            18.631598            20.718820            16.617292   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                 20.492180           23.384565           22.741578   
1                 20.770598           21.846914           24.087430   
2                 19.558451           22.068259           22.833012   
3                 20.024829           22.430380           22.405492   
4                 19.684829           23.242825           23.933544   
...                     ...                 ...                 ...   
893437            18.317562           26.345883           26.085735   
893438            19.714503           22.845933           23.283796   
893439            21.789769           23.112055           23.443923   
893440            17.320957           23.856063           22.648178   
893441            16.989700           23.481170           22.944662   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                19.956352           20.045554       1.179084       1.130519   
1                20.530784           20.293079       1.064105       1.168957   
2                20.086002           20.043214       1.104880       1.177081   
3                20.293838           20.827854       1.136516       1.130413   
4                20.487848           19.382821       1.120731       1.235955   
...                    ...                 ...            ...            ...   
893437           23.304138           23.273577       1.143444       1.127325   
893438           21.205739           19.235085       1.062779       1.220015   
893439           20.255763           20.043214       1.140421       1.182678   
893440           21.522883           21.522883       1.085827       1.063837   
893441           19.839611           20.334238       1.182774       1.140720   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            1.156924       1.120752       0.848424       0.884551   
1            1.198059       1.113785       0.939757       0.855464   
2            1.196235       1.196378       0.905535       0.849559   
3            1.218761       1.162297       0.882609       0.884633   
4            1.189528       1.123760       

In [22]:
s2 = pd.read_parquet(s2_path)
s2

long                 lat  year          species_names  \
0       75.66669580000007   32.25107380000003  2018      Bambusoideae spp.   
1       75.66669580000007   32.25107380000003  2019      Bambusoideae spp.   
2       75.66669580000007   32.25107380000003  2020      Bambusoideae spp.   
3       75.66669580000007   32.25107380000003  2021      Bambusoideae spp.   
4       75.66669580000007   32.25107380000003  2022      Bambusoideae spp.   
...                   ...                 ...   ...                    ...   
893437  78.12624330000006  31.258713300000068  2019  Rhododendron arboreum   
893438  78.12624330000006  31.258713300000068  2020  Rhododendron arboreum   
893439  78.12624330000006  31.258713300000068  2021  Rhododendron arboreum   
893440  78.12624330000006  31.258713300000068  2022  Rhododendron arboreum   
893441  78.12624330000006  31.258713300000068  2023  Rhododendron arboreum   

        B11summer  B11winter  B02summer  B02winter  B12summer  B12winter  \
0        0.289700    0.24600   0.154400    0.13630   0.235600    0.18410   
1        0.258850    0.25620   0.112800    0.14105   0.195400    0.18470   
2        0.276450    0.25100   0.144650    0.13035   0.216700    0.18535   
3        0.292600    0.17570   0.150000    0.07620   0.239950    0.10050   
4        0.307800    0.24770   0.153600    0.14870   0.234000    0.18220   
...           ...        ...        ...        ...        ...        ...   
893437   0.243600    0.16170   0.140525    0.11845   0.178200    0.13325   
893438   0.221750    0.15530   0.145950    0.15385   0.184700    0.13620   
893439   0.220675    0.16685   0.170025    0.15805   0.172975    0.13680   
893440   0.259825    0.16190   0.137625    0.13255   0.184650    0.12955   
893441   0.239050    0.16420   0.139850    0.09895   0.179750    0.13355   

        B08summer  B08winter  B04summer  B04winter  B03summer  B03winter  \
0        0.324200    0.34990    0.17640    0.14930    0.17720    0.15820   
1        0.292300    0.38535    0.13280    0.15455    0.12965    0.16150   
2        0.345250    0.35955    0.16110    0.15440    0.16560    0.15675   
3        0.319600    0.28050    0.18160    0.08550    0.17520    0.09650   
4        0.325900    0.36890    0.19020    0.15840    0.17780    0.16700   
...           ...        ...        ...        ...        ...        ...   
893437   0.346600    0.20275    0.13695    0.12440    0.14730    0.12090   
893438   0.316400    0.19710    0.14870    0.13250    0.18420    0.14450   
893439   0.337800    0.19795    0.15355    0.11840    0.16145    0.13760   
893440   0.336225    0.17620    0.13430    0.11455    0.14315    0.12110   
893441   0.332100    0.17600    0.14780    0.10340    0.16015    0.10235   

        NDVIsummer  NDVIwinter  ARVI2summer  ARVI2winter  BWDRVIsummer  \
0         0.287273    0.411741     0.156109     0.301737      0.377670   
1         0.400403    0.427497     0.288472     0.320172      0.477402   
2         0.360856    0.397863     0.242202     0.285500      0.407303   
3         0.275399    0.530091     0.142217     0.440206      0.361201   
4         0.270070    0.417861     0.135982     0.308897      0.365854   
...            ...         ...          ...          ...           ...   
893437    0.416348    0.239591     0.307127     0.100321      0.403482   
893438    0.369943    0.220117     0.252834     0.077537      0.353816   
893439    0.376100    0.157806     0.260037     0.004633      0.333711   
893440    0.423974    0.212149     0.316050     0.068215      0.409618   
893441    0.384380    0.259935     0.269725     0.124124      0.311191   

        BWDRVIwinter  CVIsummer  CVIwinter  CTVIsummer  CTVIwinter  \
0           0.439325   1.910389   2.151097    0.887284    0.954851   
1           0.465031   2.683532   2.293912    0.947785    0.963061   
2           0.466377   2.025796   2.256043    0.927653    0.947438   
3           0.570220   1.895898   2.556117    0.880560    1.014934   
4           0.43473

In [23]:
cov_df = pd.read_parquet(cov_path)
cov_df

long                 lat          species_names  \
0       75.66669580000007   32.25107380000003      Bambusoideae spp.   
1       75.66676390000003   32.25073910000003      Bambusoideae spp.   
2       75.66686810000004   32.25026770000005      Bambusoideae spp.   
3       75.66719240000003  32.249825100000066      Bambusoideae spp.   
4       75.66729090000007   32.24962370000003      Bambusoideae spp.   
...                   ...                 ...                    ...   
148902  78.12497830000007   31.25865170000003  Rhododendron arboreum   
148903  78.12522000000007  31.258880000000033  Rhododendron arboreum   
148904  78.12614830000007  31.258660000000077  Rhododendron arboreum   
148905  78.12618260000005   31.25882800000005  Rhododendron arboreum   
148906  78.12624330000006  31.258713300000068  Rhododendron arboreum   

          sand_f  elevation   cfvo_f  phh2o_f     soc_f      slope    clay_f  \
0       26.49250      352.0  13.5475    4.685   6.29375  13.422151  22.47375   
1       26.49250      350.0  13.5475    4.685   6.29375  10.147419  22.47375   
2       26.49250      349.0  13.5475    4.685   6.29375   8.110854  22.47375   
3       26.49250      351.0  13.5475    4.685   6.29375   9.826173  22.47375   
4       26.49250      348.0  13.5475    4.685   6.29375   6.125500  22.47375   
...          ...        ...      ...      ...       ...        ...       ...   
148902  36.73625     2309.0  19.9450    4.570  30.67875  54.879799  16.27250   
148903  36.73625     2292.0  19.9450    4.570  30.67875  58.185696  16.27250   
148904  36.73625     2361.0  19.9450    4.570  30.06625  30.714142  16.27250   
148905  36.73625     2350.0  19.9450    4.570  30.06625  43.765118  16.27250   
148906  36.73625     2361.0  19.9450    4.570  30.06625  30.714142  16.27250   

        nitrogen_f     cec_f    bdod_f      aspect    silt_f    ocd_f  
0           8.6365  10.74375  1.008625  254.256500  26.03375   7.0275  
1           8.6365  10.74375  1.008625  254.256454  26.03375   7.0275  
2           8.6365  10.74375  1.008625  250.076767  26.03375   7.0275  
3           8.6365  10.74375  1.008625  242.139969  26.03375   7.0275  
4           8.6365  10.74375  1.008625  243.095016  26.03375   7.0275  
...            ...       ...       ...         ...       ...      ...  
148902     11.7515  14.13875  0.968750  325.060272  21.98625  12.1675  
148903     11.7515  14.13875  0.968750  319.749359  21.98625  12.1675  
148904     11.7515  14.13875  0.966375  292.424896  21.98625  12.1675  
148905     11.7515  14.13875  0.966375  348.596802  21.98625  12.1675  
148906     11.7515  14.13875  0.966375  292.424896  21.98625  12.1675  

[148907 rows x 16 columns]

In [24]:
sent_merged_df = reduce(lambda left, right: pd.merge(left, right, on=index_cols, how='outer', validate="1:1"), [s1, s2])
sent_merged_df



long                 lat  year          species_names  \
0       75.66669580000007   32.25107380000003  2018      Bambusoideae spp.   
1       75.66669580000007   32.25107380000003  2019      Bambusoideae spp.   
2       75.66669580000007   32.25107380000003  2020      Bambusoideae spp.   
3       75.66669580000007   32.25107380000003  2021      Bambusoideae spp.   
4       75.66669580000007   32.25107380000003  2022      Bambusoideae spp.   
...                   ...                 ...   ...                    ...   
893437  78.12624330000006  31.258713300000068  2019  Rhododendron arboreum   
893438  78.12624330000006  31.258713300000068  2020  Rhododendron arboreum   
893439  78.12624330000006  31.258713300000068  2021  Rhododendron arboreum   
893440  78.12624330000006  31.258713300000068  2022  Rhododendron arboreum   
893441  78.12624330000006  31.258713300000068  2023  Rhododendron arboreum   

        vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                 22.552725            23.639657            19.700091   
1                 23.856063            23.135292            19.956352   
2                 23.268281            23.263359            21.227070   
3                 24.927604            24.248816            19.777236   
4                 24.149733            22.787536            20.453230   
...                     ...                  ...                  ...   
893437            24.899585            21.450173            22.922561   
893438            21.172713            21.895621            18.450980   
893439            20.863598            23.242825            19.138139   
893440            20.043214            20.606978            17.781513   
893441            18.631598            20.718820            16.617292   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                 20.492180           23.384565           22.741578   
1                 20.770598           21.846914           24.087430   
2                 19.558451           22.068259           22.833012   
3                 20.024829           22.430380           22.405492   
4                 19.684829           23.242825           23.933544   
...                     ...                 ...                 ...   
893437            18.317562           26.345883           26.085735   
893438            19.714503           22.845933           23.283796   
893439            21.789769           23.112055           23.443923   
893440            17.320957           23.856063           22.648178   
893441            16.989700           23.481170           22.944662   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                19.956352           20.045554       1.179084       1.130519   
1                20.530784           20.293079       1.064105       1.168957   
2                20.086002           20.043214       1.104880       1.177081   
3                20.293838           20.827854       1.136516       1.130413   
4                20.487848           19.382821       1.120731       1.235955   
...                    ...                 ...            ...            ...   
893437           23.304138           23.273577       1.143444       1.127325   
893438           21.205739           19.235085       1.062779       1.220015   
893439           20.255763           20.043214       1.140421       1.182678   
893440           21.522883           21.522883       1.085827       1.063837   
893441           19.839611           20.334238       1.182774       1.140720   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            1.156924       1.120752       0.848424       0.884551   
1            1.198059       1.113785       0.939757       0.855464   
2            1.196235       1.196378       0.905535       0.849559   
3            1.218761       1.162297       0.882609       0.884633   
4            1.189528       1.123760       

In [25]:
merged_df = reduce(lambda left, right: pd.merge(left, right,  on=["long", "lat", "species_names"], how='outer', validate="m:1"), [sent_merged_df, cov_df])
merged_df



long                 lat  year          species_names  \
0       75.66669580000007   32.25107380000003  2018      Bambusoideae spp.   
1       75.66669580000007   32.25107380000003  2019      Bambusoideae spp.   
2       75.66669580000007   32.25107380000003  2020      Bambusoideae spp.   
3       75.66669580000007   32.25107380000003  2021      Bambusoideae spp.   
4       75.66669580000007   32.25107380000003  2022      Bambusoideae spp.   
...                   ...                 ...   ...                    ...   
893437  78.12624330000006  31.258713300000068  2019  Rhododendron arboreum   
893438  78.12624330000006  31.258713300000068  2020  Rhododendron arboreum   
893439  78.12624330000006  31.258713300000068  2021  Rhododendron arboreum   
893440  78.12624330000006  31.258713300000068  2022  Rhododendron arboreum   
893441  78.12624330000006  31.258713300000068  2023  Rhododendron arboreum   

        vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                 22.552725            23.639657            19.700091   
1                 23.856063            23.135292            19.956352   
2                 23.268281            23.263359            21.227070   
3                 24.927604            24.248816            19.777236   
4                 24.149733            22.787536            20.453230   
...                     ...                  ...                  ...   
893437            24.899585            21.450173            22.922561   
893438            21.172713            21.895621            18.450980   
893439            20.863598            23.242825            19.138139   
893440            20.043214            20.606978            17.781513   
893441            18.631598            20.718820            16.617292   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                 20.492180           23.384565           22.741578   
1                 20.770598           21.846914           24.087430   
2                 19.558451           22.068259           22.833012   
3                 20.024829           22.430380           22.405492   
4                 19.684829           23.242825           23.933544   
...                     ...                 ...                 ...   
893437            18.317562           26.345883           26.085735   
893438            19.714503           22.845933           23.283796   
893439            21.789769           23.112055           23.443923   
893440            17.320957           23.856063           22.648178   
893441            16.989700           23.481170           22.944662   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                19.956352           20.045554       1.179084       1.130519   
1                20.530784           20.293079       1.064105       1.168957   
2                20.086002           20.043214       1.104880       1.177081   
3                20.293838           20.827854       1.136516       1.130413   
4                20.487848           19.382821       1.120731       1.235955   
...                    ...                 ...            ...            ...   
893437           23.304138           23.273577       1.143444       1.127325   
893438           21.205739           19.235085       1.062779       1.220015   
893439           20.255763           20.043214       1.140421       1.182678   
893440           21.522883           21.522883       1.085827       1.063837   
893441           19.839611           20.334238       1.182774       1.140720   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            1.156924       1.120752       0.848424       0.884551   
1            1.198059       1.113785       0.939757       0.855464   
2            1.196235       1.196378       0.905535       0.849559   
3            1.218761       1.162297       0.882609       0.884633   
4            1.189528       1.123760       

### Species Codes and JSON File

In [26]:
if not species_dict_path.exists():

    #exporting species dictionary as json
    cats = pd.Categorical(merged_df['species_names'])
    species_dict = dict(zip(range(len(cats.categories)), cats.categories))

    with open(species_dict_path, 'w') as json_file:
        json.dump(species_dict, json_file, indent=4)

In [27]:
merged_df["species"] = pd.Categorical(merged_df['species_names']).codes
merged_df.drop('species_names', axis=1, inplace=True)
merged_df

long                 lat  year  vv_Descendingsummer  \
0       75.66669580000007   32.25107380000003  2018            22.552725   
1       75.66669580000007   32.25107380000003  2019            23.856063   
2       75.66669580000007   32.25107380000003  2020            23.268281   
3       75.66669580000007   32.25107380000003  2021            24.927604   
4       75.66669580000007   32.25107380000003  2022            24.149733   
...                   ...                 ...   ...                  ...   
893437  78.12624330000006  31.258713300000068  2019            24.899585   
893438  78.12624330000006  31.258713300000068  2020            21.172713   
893439  78.12624330000006  31.258713300000068  2021            20.863598   
893440  78.12624330000006  31.258713300000068  2022            20.043214   
893441  78.12624330000006  31.258713300000068  2023            18.631598   

        vv_Descendingwinter  vh_Descendingsummer  vh_Descendingwinter  \
0                 23.639657            19.700091            20.492180   
1                 23.135292            19.956352            20.770598   
2                 23.263359            21.227070            19.558451   
3                 24.248816            19.777236            20.024829   
4                 22.787536            20.453230            19.684829   
...                     ...                  ...                  ...   
893437            21.450173            22.922561            18.317562   
893438            21.895621            18.450980            19.714503   
893439            23.242825            19.138139            21.789769   
893440            20.606978            17.781513            17.320957   
893441            20.718820            16.617292            16.989700   

        vv_Ascendingsummer  vv_Ascendingwinter  vh_Ascendingsummer  \
0                23.384565           22.741578           19.956352   
1                21.846914           24.087430           20.530784   
2                22.068259           22.833012           20.086002   
3                22.430380           22.405492           20.293838   
4                23.242825           23.933544           20.487848   
...                    ...                 ...                 ...   
893437           26.345883           26.085735           23.304138   
893438           22.845933           23.283796           21.205739   
893439           23.112055           23.443923           20.255763   
893440           23.856063           22.648178           21.522883   
893441           23.481170           22.944662           19.839611   

        vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  VV_VH_Dsummer  \
0                20.045554       1.179084       1.130519       1.156924   
1                20.293079       1.064105       1.168957       1.198059   
2                20.043214       1.104880       1.177081       1.196235   
3                20.827854       1.136516       1.130413       1.218761   
4                19.382821       1.120731       1.235955       1.189528   
...                    ...            ...            ...            ...   
893437           23.273577       1.143444       1.127325       1.093749   
893438           19.235085       1.062779       1.220015       1.130216   
893439           20.043214       1.140421       1.182678       1.123455   
893440           21.522883       1.085827       1.063837       1.100190   
893441           20.334238       1.182774       1.140720       1.045365   

        VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  VH_VV_Dsummer  \
0            1.120752       0.848424       0.884551       0.865649   
1            1.113785       0.939757       0.855464       0.834683   
2            1.196378       0.905535       0.849559       0.836198   
3            1.162297       0.882609       0.884633       0.820505   
4            1.123760       0.892275       0.809453       0.840669   
...               ...            ...            ...            ...   

In [28]:
merged_df['species'].unique()

array([ 1,  4, 11,  6,  5,  7,  8,  3, 13,  2,  0, 10, 12,  9], dtype=int8)

### Normalizing

In [29]:
df_index_cols = merged_df[['long', 'lat', 'year','species']]
df_x = merged_df.drop(['long', 'lat', 'year','species'], axis=1)

df_x

vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                 22.552725            23.639657            19.700091   
1                 23.856063            23.135292            19.956352   
2                 23.268281            23.263359            21.227070   
3                 24.927604            24.248816            19.777236   
4                 24.149733            22.787536            20.453230   
...                     ...                  ...                  ...   
893437            24.899585            21.450173            22.922561   
893438            21.172713            21.895621            18.450980   
893439            20.863598            23.242825            19.138139   
893440            20.043214            20.606978            17.781513   
893441            18.631598            20.718820            16.617292   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                 20.492180           23.384565           22.741578   
1                 20.770598           21.846914           24.087430   
2                 19.558451           22.068259           22.833012   
3                 20.024829           22.430380           22.405492   
4                 19.684829           23.242825           23.933544   
...                     ...                 ...                 ...   
893437            18.317562           26.345883           26.085735   
893438            19.714503           22.845933           23.283796   
893439            21.789769           23.112055           23.443923   
893440            17.320957           23.856063           22.648178   
893441            16.989700           23.481170           22.944662   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                19.956352           20.045554       1.179084       1.130519   
1                20.530784           20.293079       1.064105       1.168957   
2                20.086002           20.043214       1.104880       1.177081   
3                20.293838           20.827854       1.136516       1.130413   
4                20.487848           19.382821       1.120731       1.235955   
...                    ...                 ...            ...            ...   
893437           23.304138           23.273577       1.143444       1.127325   
893438           21.205739           19.235085       1.062779       1.220015   
893439           20.255763           20.043214       1.140421       1.182678   
893440           21.522883           21.522883       1.085827       1.063837   
893441           19.839611           20.334238       1.182774       1.140720   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            1.156924       1.120752       0.848424       0.884551   
1            1.198059       1.113785       0.939757       0.855464   
2            1.196235       1.196378       0.905535       0.849559   
3            1.218761       1.162297       0.882609       0.884633   
4            1.189528       1.123760       0.892275       0.809453   
...               ...            ...            ...            ...   
893437       1.093749       1.171387       0.874551       0.897222   
893438       1.130216       1.122397       0.942758       0.819662   
893439       1.123455       1.137903       0.877059       0.848131   
893440       1.100190       1.147706       0.920957       0.940453   
893441       1.045365       1.103533       0.846055       0.876639   

        VH_VV_Dsummer  VH_VV_Dwinter  SAR_NDVI_Asummer  SAR_NDVI_Awinter  \
0            0.865649       0.892258         -0.082086         -0.061261   
1            0.834683       0.897885         -0.031057         -0.077898   
2            0.836198       0.835856         -0.049694         -0.081339   
3            0.820505       0.860365         -0.063077         -0.061215   
4            0.840669       0.889870         -0.056929         -0.105405   
...               ..

In [30]:
min_max_scaler = MinMaxScaler()

df_scaled = pd.DataFrame(min_max_scaler.fit_transform(df_x.values), columns=df_x.columns)
df_scaled

vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                  0.396606             0.471055             0.416042   
1                  0.460550             0.446541             0.429824   
2                  0.431712             0.452766             0.498169   
3                  0.513121             0.500663             0.420191   
4                  0.474957             0.429639             0.456548   
...                     ...                  ...                  ...   
893437             0.511746             0.364638             0.589358   
893438             0.328901             0.386289             0.348860   
893439             0.313736             0.451768             0.385818   
893440             0.273487             0.323656             0.312853   
893441             0.204231             0.329092             0.250237   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                  0.470285            0.451115            0.404824   
1                  0.487266            0.366455            0.478185   
2                  0.413335            0.378641            0.409808   
3                  0.441780            0.398579            0.386505   
4                  0.421043            0.443311            0.469797   
...                     ...                 ...                 ...   
893437             0.337652            0.614159            0.587110   
893438             0.422853            0.421459            0.434380   
893439             0.549427            0.436111            0.443108   
893440             0.276867            0.477074            0.399733   
893441             0.256663            0.456433            0.415894   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                 0.341066            0.350320       0.338847       0.269886   
1                 0.370159            0.362405       0.183898       0.322430   
2                 0.347632            0.350206       0.238847       0.333536   
3                 0.358158            0.388514       0.281481       0.269742   
4                 0.367985            0.317964       0.260208       0.414014   
...                    ...                 ...            ...            ...   
893437            0.510621            0.507920       0.290817       0.265521   
893438            0.404344            0.310751       0.182112       0.392224   
893439            0.356230            0.350206       0.286744       0.341186   
893440            0.420406            0.422447       0.213171       0.178735   
893441            0.335153            0.364414       0.343820       0.283832   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            0.336448       0.307829       0.519535       0.602601   
1            0.383513       0.300665       0.709946       0.540842   
2            0.381426       0.385602       0.638601       0.528305   
3            0.407200       0.350554       0.590805       0.602773   
4            0.373752       0.310922       0.610956       0.443153   
...               ...            ...            ...            ...   
893437       0.264165       0.359902       0.574005       0.629503   
893438       0.305890       0.309521       0.716203       0.464829   
893439       0.298154       0.325467       0.579233       0.525273   
893440       0.271534       0.335549       0.670753       0.721291   
893441       0.208805       0.290121       0.514597       0.585801   

        VH_VV_Dsummer  VH_VV_Dwinter  SAR_NDVI_Asummer  SAR_NDVI_Awinter  \
0            0.497106       0.506712          0.584585          0.662452   
1            0.444008       0.515246          0.761503          0.603882   
2            0.446605       0.421169          0.696887          0.591767   
3            0.419696       0.458341          0.650487          0.662615   
4            0.454273       0.503090          0.671804          0.507039   
...               ..

### Model Data Frame

In [31]:
model_df = pd.concat([df_index_cols, df_scaled], axis=1)
model_df

long                 lat  year  species  \
0       75.66669580000007   32.25107380000003  2018        1   
1       75.66669580000007   32.25107380000003  2019        1   
2       75.66669580000007   32.25107380000003  2020        1   
3       75.66669580000007   32.25107380000003  2021        1   
4       75.66669580000007   32.25107380000003  2022        1   
...                   ...                 ...   ...      ...   
893437  78.12624330000006  31.258713300000068  2019       10   
893438  78.12624330000006  31.258713300000068  2020       10   
893439  78.12624330000006  31.258713300000068  2021       10   
893440  78.12624330000006  31.258713300000068  2022       10   
893441  78.12624330000006  31.258713300000068  2023       10   

        vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                  0.396606             0.471055             0.416042   
1                  0.460550             0.446541             0.429824   
2                  0.431712             0.452766             0.498169   
3                  0.513121             0.500663             0.420191   
4                  0.474957             0.429639             0.456548   
...                     ...                  ...                  ...   
893437             0.511746             0.364638             0.589358   
893438             0.328901             0.386289             0.348860   
893439             0.313736             0.451768             0.385818   
893440             0.273487             0.323656             0.312853   
893441             0.204231             0.329092             0.250237   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                  0.470285            0.451115            0.404824   
1                  0.487266            0.366455            0.478185   
2                  0.413335            0.378641            0.409808   
3                  0.441780            0.398579            0.386505   
4                  0.421043            0.443311            0.469797   
...                     ...                 ...                 ...   
893437             0.337652            0.614159            0.587110   
893438             0.422853            0.421459            0.434380   
893439             0.549427            0.436111            0.443108   
893440             0.276867            0.477074            0.399733   
893441             0.256663            0.456433            0.415894   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                 0.341066            0.350320       0.338847       0.269886   
1                 0.370159            0.362405       0.183898       0.322430   
2                 0.347632            0.350206       0.238847       0.333536   
3                 0.358158            0.388514       0.281481       0.269742   
4                 0.367985            0.317964       0.260208       0.414014   
...                    ...                 ...            ...            ...   
893437            0.510621            0.507920       0.290817       0.265521   
893438            0.404344            0.310751       0.182112       0.392224   
893439            0.356230            0.350206       0.286744       0.341186   
893440            0.420406            0.422447       0.213171       0.178735   
893441            0.335153            0.364414       0.343820       0.283832   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            0.336448       0.307829       0.519535       0.602601   
1            0.383513       0.300665       0.709946       0.540842   
2            0.381426       0.385602       0.638601       0.528305   
3            0.407200       0.350554       0.590805       0.602773   
4            0.373752       0.310922       0.610956       0.443153   
...               ...            ...            ...            ...   
893437       0.264165       0.359902       0.574005       0.629503   
8

In [32]:
for col in model_df.columns:
    print(pd.Series(model_df[col].isna().value_counts()))

long
False    893442
Name: count, dtype: int64
lat
False    893442
Name: count, dtype: int64
year
False    893442
Name: count, dtype: int64
species
False    893442
Name: count, dtype: int64
vv_Descendingsummer
False    893442
Name: count, dtype: int64
vv_Descendingwinter
False    893442
Name: count, dtype: int64
vh_Descendingsummer
False    893442
Name: count, dtype: int64
vh_Descendingwinter
False    893442
Name: count, dtype: int64
vv_Ascendingsummer
False    893442
Name: count, dtype: int64
vv_Ascendingwinter
False    893442
Name: count, dtype: int64
vh_Ascendingsummer
False    893442
Name: count, dtype: int64
vh_Ascendingwinter
False    893442
Name: count, dtype: int64
VV_VH_Asummer
False    893442
Name: count, dtype: int64
VV_VH_Awinter
False    893442
Name: count, dtype: int64
VV_VH_Dsummer
False    893442
Name: count, dtype: int64
VV_VH_Dwinter
False    893442
Name: count, dtype: int64
VH_VV_Asummer
False    893442
Name: count, dtype: int64
VH_VV_Awinter
False    893442
Name: co

### Balanced Panel of Species

For seven species, (13,3,1,0,5,10,6) which have locations less than 5000, we have taken all the locations, and for the remaining 7, we have randomly sampled 5000 locations brininging the total observations to 64,147. This was the actual data frame with 532 features. But for ease of modelling we are restructuring the model frame a little but we will retain the sampling of 5000 locations per species in the actual frame. 

We merge the lat longs of the actual data frame (64,147 * 532) with our existing frame and then retain only 5000 locations per specie in case where we have more than 5000 locations in the data and if the data has locations less 5000, we keep all.

Here after grouping by Lat and Long, the number of rows equal the 148,907 which is our actual number of unique locations. After merge we expect this to be 64,147

In [33]:
model_df.groupby(['lat', 'long'])['species'].count().reset_index()

lat               long  species
0        30.38513370000004  77.57655330000006        6
1       30.387641900000062  77.58138280000009        6
2       30.389346600000067  77.56340700000004        6
3       30.389401600000042  77.56335570000005        6
4       30.389493300000026  77.56316670000007        6
...                    ...                ...      ...
148902   32.94831170000003  76.45642000000004        6
148903   33.08003080000003  76.40363490000004        6
148904   33.08007490000006  76.40368610000007        6
148905  33.080129500000055  76.40363320000006        6
148906  33.080214400000045  76.40355620000008        6

[148907 rows x 3 columns]

In [34]:
actual_df = pd.read_csv(actual_model_path,  dtype={"long": str, "lat": str})
actual_df

long                 lat  species  \
0      75.87420370000007   32.83771860000007        0   
1      75.87430770000003  32.839617700000076        0   
2      75.87460430000004   32.83970240000008        0   
3      75.87476720000006  32.840192100000024        0   
4      75.87597120000004  32.840583700000025        0   
...                  ...                 ...      ...   
64142  78.05004830000007  31.207106700000054       13   
64143  78.05013670000005  31.206146700000033       13   
64144  78.05047500000006  31.206276700000046       13   
64145  78.05075500000004   31.20620260000004       13   
64146  78.09057170000005  31.195590700000025       13   

       vv_Descendingsummer2018  vv_Descendingsummer2019  \
0                     0.608198                 0.810302   
1                     0.606787                 0.649423   
2                     0.578037                 0.643389   
3                     0.554325                 0.732400   
4                     0.315383                 0.342842   
...                        ...                      ...   
64142                 0.167616                 0.474936   
64143                 0.247722                 0.407294   
64144                 0.252469                 0.493193   
64145                 0.346617                 0.560047   
64146                 0.340399                 0.396500   

       vv_Descendingsummer2020  vv_Descendingsummer2021  \
0                     0.264052                 0.277363   
1                     0.444631                 0.436618   
2                     0.292008                 0.382898   
3                     0.379040                 0.412763   
4                     0.322986                 0.328509   
...                        ...                      ...   
64142                 0.410658                 0.543295   
64143                 0.430835                 0.509509   
64144                 0.469372                 0.489009   
64145                 0.481431                 0.569621   
64146                 0.295365                 0.377266   

       vv_Descendingsummer2022  vv_Descendingsummer2023  \
0                     0.243221                 0.499439   
1                     0.391040                 0.301967   
2                     0.362196                 0.312260   
3                     0.424432                 0.364360   
4                     0.310343                 0.443069   
...                        ...                      ...   
64142                 0.499446                 0.405003   
64143                 0.429288                 0.380439   
64144                 0.494462                 0.380656   
64145                 0.515031                 0.406095   
64146                 0.345183                 0.370128   

       vv_Descendingwinter2018  vv_Descendingwinter2019  \
0                     0.756396                 0.722973   
1                     0.677898                 0.599708   
2                     0.692859                 0.656819   
3                     0.633982                 0.654292   
4                     0.321656                 0.328801   
...                        ...                      ...   
64142                 0.469956                 0.399973   
64143                 0.373685                 0.281528   
64144                 0.478518                 0.386776   
64145                 0.497513                 0.353148   
64146                 0.531958                 0.352174   

       vv_Descendingwinter2020  vv_Descendingwinter2021  \
0                     0.286786                 0.241322   
1                     0.450024                 0.463995   
2                     0.325859                 0.361550   
3                     0.396818                 0.444452   
4                     0.387857                 0.356554   
...                        ...                      ...   
64142                 0.523307                 0.664063   
64143                 0.486

In [35]:
species_sampled = pd.merge(model_df, actual_df[['long', 'lat']], on=['long', 'lat'], how='inner', validate='m:1')
species_sampled.groupby(['lat', 'long'])['species'].count().reset_index()

lat               long  species
0       30.38513370000004  77.57655330000006        6
1      30.387641900000062  77.58138280000009        6
2      30.389401600000042  77.56335570000005        6
3      30.389493300000026  77.56316670000007        6
4       30.38954030000008  77.56328840000003        6
...                   ...                ...      ...
64142   32.94321140000005  76.46184580000005        6
64143  32.943245700000034  76.17807740000006        6
64144   32.94332000000003  76.17807380000005        6
64145  32.948300000000074  76.45644330000005        6
64146  33.080214400000045  76.40355620000008        6

[64147 rows x 3 columns]

The number of rows above is now 64,147 during groupby using Lat and Long.

### Checking for species count
The following process ensures that we have the same number per species after processing.

In [36]:
unique_loc = species_sampled.groupby(['lat', 'long'])['species'].count().reset_index().drop('species', axis=1)
unique_loc.merge(actual_df[['long', 'lat', 'species']], on=['long', 'lat'], how='outer')['species'].value_counts()


species
4     5000
11    5000
7     5000
12    5000
2     5000
8     5000
9     5000
6     4834
10    4765
5     4737
0     4206
1     3939
3     3524
13    3142
Name: count, dtype: int64

### Frame to be considered for modelling

In [37]:
species_sampled


long                 lat  year  species  \
0       75.66669580000007   32.25107380000003  2018        1   
1       75.66669580000007   32.25107380000003  2019        1   
2       75.66669580000007   32.25107380000003  2020        1   
3       75.66669580000007   32.25107380000003  2021        1   
4       75.66669580000007   32.25107380000003  2022        1   
...                   ...                 ...   ...      ...   
384877  78.12624330000006  31.258713300000068  2019       10   
384878  78.12624330000006  31.258713300000068  2020       10   
384879  78.12624330000006  31.258713300000068  2021       10   
384880  78.12624330000006  31.258713300000068  2022       10   
384881  78.12624330000006  31.258713300000068  2023       10   

        vv_Descendingsummer  vv_Descendingwinter  vh_Descendingsummer  \
0                  0.396606             0.471055             0.416042   
1                  0.460550             0.446541             0.429824   
2                  0.431712             0.452766             0.498169   
3                  0.513121             0.500663             0.420191   
4                  0.474957             0.429639             0.456548   
...                     ...                  ...                  ...   
384877             0.511746             0.364638             0.589358   
384878             0.328901             0.386289             0.348860   
384879             0.313736             0.451768             0.385818   
384880             0.273487             0.323656             0.312853   
384881             0.204231             0.329092             0.250237   

        vh_Descendingwinter  vv_Ascendingsummer  vv_Ascendingwinter  \
0                  0.470285            0.451115            0.404824   
1                  0.487266            0.366455            0.478185   
2                  0.413335            0.378641            0.409808   
3                  0.441780            0.398579            0.386505   
4                  0.421043            0.443311            0.469797   
...                     ...                 ...                 ...   
384877             0.337652            0.614159            0.587110   
384878             0.422853            0.421459            0.434380   
384879             0.549427            0.436111            0.443108   
384880             0.276867            0.477074            0.399733   
384881             0.256663            0.456433            0.415894   

        vh_Ascendingsummer  vh_Ascendingwinter  VV_VH_Asummer  VV_VH_Awinter  \
0                 0.341066            0.350320       0.338847       0.269886   
1                 0.370159            0.362405       0.183898       0.322430   
2                 0.347632            0.350206       0.238847       0.333536   
3                 0.358158            0.388514       0.281481       0.269742   
4                 0.367985            0.317964       0.260208       0.414014   
...                    ...                 ...            ...            ...   
384877            0.510621            0.507920       0.290817       0.265521   
384878            0.404344            0.310751       0.182112       0.392224   
384879            0.356230            0.350206       0.286744       0.341186   
384880            0.420406            0.422447       0.213171       0.178735   
384881            0.335153            0.364414       0.343820       0.283832   

        VV_VH_Dsummer  VV_VH_Dwinter  VH_VV_Asummer  VH_VV_Awinter  \
0            0.336448       0.307829       0.519535       0.602601   
1            0.383513       0.300665       0.709946       0.540842   
2            0.381426       0.385602       0.638601       0.528305   
3            0.407200       0.350554       0.590805       0.602773   
4            0.373752       0.310922       0.610956       0.443153   
...               ...            ...            ...            ...   
384877       0.264165       0.359902       0.574005       0.629503   
3

### Exporting File

In [38]:
species_sampled.to_parquet(model_df_path, index=False)
# model_df.to_parquet(species_model_df_path, index=False)

In [39]:
# species_sampled[species_sampled["species"].isin([10,8])]